### ISSUE TITLE
Jumped sessions not being properly terminated when target server shutdown.

### ISSUE THREAD

**Issue by jabbera** [ronf/asyncssh#471]
Hi!

Is there any way to set SO_KEEPALIVE here without a PR or copying the forward_connection method into my code? There isn't a factory parameter on this method allowing me to override the SSHForwarder type. 

https://github.com/ronf/asyncssh/blob/37d0afc7f40382b588f287d2346de43094d20262/asyncssh/forward.py#L96

As for my reason, if you use asyncssh as a jump host and yank the network cable (or shutdown a k8s pod) on the forward target without a clean shutdown the ssh client just hangs forever. The ssh keepalives between the client and asyncssh are good, but the target is deader then dead and can't respond at all.

Thanks!
Mike

**Comment by ronf** [comment]
If you're looking for a way to make a tunnel request from an SSH client request that the server set SO_KEEPALIVE on the remote tunnel connection, there's no way to do this in the SSH protocol. The best you'd be able to do here would be to unconditionally set SO_KEEPALIVE on all local tunneled connections which are opened by AsyncSSH, and that will only work if AsyncSSH is actually the server opening those tunnel connections. If the tunnels are being opened by some other SSH implementation, you'd have to make the change there.

That said, the place you've identified above would probably be the right place to do this if you wanted it to apply unconditionally to all local TCP tunnel connections. The tricky part would be passing in option information from the SSH server level to the SSHForwarder, as SSHForwarder objects are generally created automatically based on requests coming in on the SSH connection. So, there's no place for application code to directly pass in whether to enable keepalive or not.

One other thing to keep in mind is that the TCP keepalive mechanism is notoriously slow to recognize dead connections. If there's no data being sent on the connection it can take something like 2+ hours by default to actually do the detection. It's much quick if data is being sent, but in that case the failure will be detected even without enabling SO_KEEPALIVE on the socket.

**Comment by jabbera** [comment]
Thanks for the response. I have the keepalive set to 30 seconds on the node that's running this. I'm open to other solutions to my issue (ssh client jumped via asyncssh hanging indefinitely when target stops responding to network traffic) but keep alives seemed to be the easy choice. Would you consider a PR that allows an end user to override the class used when forward_connection is called?

**Comment by ronf** [comment]
A host not responding shouldn't cause AsyncSSH to hang. In fact, if anything, writing data that doesn't get an acknowledgement should pretty quickly lead to the connection failing even without TCP keepalive enabled. If you're seeing a hang, you may have uncovered a bug of some kind. Even if you don't write any data and are just reading, that should only cause the task doing the read to hang, but it shouldn't prevent any other tasks you might be running in parallel from still making progress.

Regarding passing in a factory method, you can pretty much already do that today in the `connection_requested` callback of SSHServerConnection. See the last example under https://asyncssh.readthedocs.io/en/latest/#port-forwarding where that function calls forward_connection instead of just returning `True` or `False`. However, there are multiple layers of subclasses under SSHForwarder to handle different types of tunnels, and those classes don't expose the underlying socket that gets opened up.

I'd need to do a bit more exploration to see how much of a change this would introduce, but the simplest thing I can think of here would be to allow the existing "tcp_keepalive" option in SSHServerConnectionOptions to apply to both SSH connections and tunneled connections. I'd still need to find a way to pass that option value through to the SSHForwarder, but that would minimize changes needed in application code.

If you needed to make a different decision on the keepalive for SSH connections vs. tunnel connections, I could add an extra option like `tunnel_keepalive` here, but I'm hesitant to do that. It's a bit of an edge case, and there's already potential confusion between the SSH keepalive options and this TCP keepalive option. I don't think I want to introduce a third one.

Would use the existing `tcp_keepalive` option work for you?

**Comment by jabbera** [comment]
Keepalives did not fix my problem as you mentioned. 

Turns out there is a bug (I think). 

If I have a tunneled connection through asyncssh setup like so:

```
ssh -AJ me@localhost:8024 me@me.me.tld
```
using the following server:

```
class NotebookSSHServer(asyncssh.SSHServer):
    async def begin_auth(self, username: str) -> bool:
        return False

    def connection_made(self, conn: asyncssh.SSHServerConnection) -> None:
        self.__conn = conn

    def connection_requested(self, dest_host: str, dest_port: int, orig_host: str, orig_port: int):
        return self.__conn.forward_connection(dest_host, dest_port)
```

Then I terminate the SSH server I end up with this in my log and ssh keepalives between my client and asyncssh continue forever and I'm locked up:

```
DEBUG:asyncssh:[conn=0, chan=0] Sending 100 data bytes
DEBUG:asyncssh:[conn=0, chan=0] Sending EOF
DEBUG:asyncssh:[conn=0, chan=0] Received EOF
DEBUG:asyncssh:[conn=0] Sending keepalive request
DEBUG:asyncssh:[conn=0] Got keepalive response
DEBUG:asyncssh:[conn=0] Sending keepalive request
DEBUG:asyncssh:[conn=0] Got keepalive response
DEBUG:asyncssh:[conn=0] Sending keepalive request
DEBUG:asyncssh:[conn=0] Got keepalive response
```

In debugging the code this line returns False:

https://github.com/ronf/asyncssh/blob/37d0afc7f40382b588f287d2346de43094d20262/asyncssh/forward.py#L127

I end up in asyncio.selector_events:

![image](https://user-images.githubusercontent.com/301208/161409116-3d49be6f-a02d-4c7c-9f9d-dc7efd7b9d57.png)

Then nothing else ever happens.

Full gist of log here: https://gist.github.com/jabbera/816042dc76f19540e65222f10c29f3c8
Full source code used for repro here: https://gist.github.com/jabbera/d3678346873778409f040bf96ed11642

**Comment by ronf** [comment]
When you say you "terminate the SSH server", how were you doing this? Also, by "locked up", do you just mean that the AsyncSSH server remains blocked in run_forever() rather than exiting?

Looking at the example code, I see it starting an SSH listener, but I don't see anything which is stopping that listener. So, even after the forwarded tunnel connection is closed, the SSH control connection will remain open, and so will the SSH listener which is accepting new connections. Nothing on the server side is doing anything to shut down either of these.

**Comment by jabbera** [comment]
Let me lay out my use case as background: I'm using asyncssh as a jump host for connecting to on the fly allocated compute in a Kubernetes cluster. In connection_requested I'm checking for existing allocated compute for the authed user and if it doesn't exist spinning it up. 

All of this works great. Where I am hitting this issue is: I have a traditional openssh client connected via asyncssh as a jump host to the k8s compute running openssh sshd. If I shutdown the k8s pod that's running sshd the ssh client that's tunneled through asyncssh will never detect that the underlying target host isn't reachable. I'll try to type something and my input goes no where (obviously). Ssh keepalives keep flowing between the client and asyncssh. I have to force terminate the client. I'd expect asyncssh to eventually let the openssh client know that it's connected tunnel is gone and to disconnect it. This doesn't seem to be happening.

**Comment by jabbera** [comment]
I managed to work around my issue with a pretty ugly hack so I either don't have the extensibility point I need, it's a bug, or I'm doing this wrong. connection_lost on SSHForwarder never seems to be called if the server is terminated. 

I copied forward_connection and accessed your private _loop substituting a custom SSHForwarder that closes the connection when it gets an EOF.

```
class MySSHForwarder(asyncssh.SSHForwarder):
    def __init__(self, peer: Optional["SSHForwarder"] = None, conn):
        super().__init__(peer)
        self._conn = conn

    def write_eof(self) -> None:
        super().write_eof()
        if self._conn:
            self._conn.close()

    # This is never called if I don't add the close above
    def connection_lost(self, exc: Optional[Exception]) -> None:
        super().connection_lost(exc)


class NotebookSSHServer(asyncssh.SSHServer):
    async def forward_connection(self, dest_host: str, dest_port: int) -> asyncssh.SSHForwarder:
        try:
            _, peer = await self.__conn._loop.create_connection(asyncssh.SSHForwarder, dest_host, dest_port)

            logging.info("  Forwarding TCP connection to %s", (dest_host, dest_port))
        except OSError as exc:
            raise asyncssh.ChannelOpenError(asyncssh.constants.OPEN_CONNECT_FAILED, str(exc)) from None

        return MySSHForwarder(peer, self.__conn)

    async def begin_auth(self, username: str) -> bool:
        return False

    def connection_made(self, conn: asyncssh.SSHServerConnection) -> None:
        self.__conn = conn

    def connection_requested(self, dest_host: str, dest_port: int, orig_host: str, orig_port: int):
        return self.forward_connection(dest_host, dest_port)
```

**Comment by ronf** [comment]
Thanks for the added info. It seems to me that something's not quite right here in the handling of "half-open" connections, where one side has sent EOF but it is leaving the connection open to allow the other side to continue to write data, and thus it is holding off on calling close until both sides have sent EOF.

That said, your debugging output shows that you have both sent & received EOF, so I'd expect after that for the tunnel connection to close, and once that happened it should cause the SSH control connection opened via the jump host to
close. Since both OpenSSH and AsyncSSH are involved here, though, and we're dealing with two different SSH "connections" plus a tunnel, there's lots of room for things to go wrong.

Can you collect debug logs on the SSH client? Also, can you confirm the arguments you are giving it (or the config file settings used) to open the connection through the AsyncSSH jump host? I'm starting to think that the bug might be in OpenSSH, since AsyncSSH appears to be properly closing the tunnel, but the OpenSSH client isn't reacting when the tunnel is closed.

Regarding your workaround, forcing a close like this for the whole connection when the tunnel closes might work for your use case, but it would mean that you couldn't have multiple tunnels open at once. As soon as one tunnel on a jump host connection closed, all the other tunnels on that same connection would be killed.

Perhaps just triggering an explicit `close()` on the tunnel connection inside of `conn` might be enough here. Could you try `self.close()` in your MySSHForwarder's `write_eof()` call rather than `self._conn.close()`? Doing that unconditionally is still not quite right, but if that works I can look into changing how eof_received() works to force a close once both sides have sent EOF.

**Comment by ronf** [comment]
Also, could you try the following change?

```patch
diff --git a/asyncssh/forward.py b/asyncssh/forward.py
index 784db1d..872389a 100644
--- a/asyncssh/forward.py
+++ b/asyncssh/forward.py
@@ -126,7 +126,7 @@ class SSHForwarder(asyncio.BaseProtocol):

             return not self._peer.was_eof_received()
         else:
-            return True
+            return False

     def pause_writing(self) -> None:
         """Pause writing by asking peer to pause reading"""
```

This forces a close whenever there is no "peer" associated with the forwarded connection. I'm not sure this is the right thing, as I think this would only happen when a connection was first being set up and fully closing a connection before it is fully opened may cause problems. However, I'm curious if this changes the behavior at all in your case.

**Comment by jabbera** [comment]
> Could you try `self.close()` in your MySSHForwarder's `write_eof()` call rather than `self._conn.close()`

This change resolves the issue.

> Also, could you try the following change? (subsequent patch)

This does not resolve the issue


Log from both async ssh and open ssh on the issue:

Command used (no config file): ssh -v -J account@localhost:8024 account@target

Openssh: https://gist.github.com/jabbera/7548059b578686ae3332cbf0549b8e4e
Async SSH: https://gist.github.com/jabbera/b392320843d09ce62d2e3ea21aa548a6

**Comment by jabbera** [comment]
I've been applying this diff after my pip install and I'm working well. I dunno if it's the right thing to do but it solves my issue! Thanks for all of your help:

```
--- ./.venv/lib/python3.10/site-packages/asyncssh/forward.py	2022-04-03 18:55:04.860000000 -0400
+++ ./forward.py	2022-04-03 18:57:42.350000000 -0400
@@ -68,6 +68,7 @@
         except OSError: # pragma: no cover
             pass
 
+        self.close()
     def was_eof_received(self) -> bool:
         """Return whether end of file has been received or not"""

```

**Comment by ronf** [comment]
Thanks. I did some additional exploration on this tonight, and I think I was able to reproduce the issue here.

From what I can see, both AsyncSSH and OpenSSH acting as the jump server have similar behavior. If the connection to the server goes away and the client isn't actively sending any data over the tunnel to the remote host, it won't detect connection failures (at least not quickly). As soon as data is sent, though, it only takes it a minute or so before the client gets a connection failure.

A better fix for this problem might be to enable keepalive on the client (ServerAliveInterval). That will end up doing keepalives on both the connection to the jump host AND the tunneled connection to the target host. With a 5 second ServerAliveInterval and the default of 3 for ServerAliveCountMax, the connection failure will be detected in about 15 seconds  (5*3), even when there's no real data in flight. Here's the command I used for the test:

`ssh -J jumphost:8022 -o ServerAliveInterval=5 target`

Any attempt to forcibly close the connection in the jump host will only work if you do that after only one side of the connection is closed, potentially causing data still in flight from the other side to get dropped. So, while that' working for you here, that will probably cause problems in other cases, potentially dropping output before it is all received if a client does a half-close on its side of the connection after writing all the data it plans on sending.

**Comment by jabbera** [comment]
I can't depend on all of my users to set themselves up properly unfortunately. I'm pretty sure for my specific use case this will be fine as once a session is established it's not going away unless shutdowns are imminent. Thanks for your help!

**Comment by ronf** [comment]
The main concern would be a client that sends EOF before reading all the data they expect to get back. That's probably not an issue too much for interactive shell sessions, so I think you'd be ok there. However, I could see this being a problem if you ever had client which execute commands via SSH rather than starting a shell. If stdin is pointing at a file (or /dev/null), this early close() could cause AsyncSSH to close the channel before all the output going to stdout/stderr was forwarded to the client.

**Comment by jabbera** [comment]
If anyone ever stumbles across this issue it's resolved as of 2.21.1. The patch above is no longer needed.
